### PR TITLE
Single partition recovery

### PR DESCRIPTION
**Pull request by kbatuigas**
[Deploy preview](https://deploy-preview-236--redpanda-docs-preview.netlify.app/current/manage/cluster-maintenance/partition-recovery/)

- Explain when and how to carry out partition recovery with the Admin API. (API doc updates were done in #153 ).
- This is also an accompanying doc to the nodewise recovery updates (separate PR) - an update will be needed to cross link the two docs.
- 
Questions for SME are added as comments in this PR

Closes https://github.com/redpanda-data/documentation-private/issues/2105

### CODE REVIEW

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 1]
```diff
@@ -0,0 +1,30 @@
+= Partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can arise when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Partition recovery is carried out by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. However, this allows some potential data loss on the partition if the best available replica is out of sync, resulting in an unclean leader election. Partition recovery should only be used when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
+
+If you want to instead permanently remove failed brokers from a cluster, especially in situations where no surviving partition replicas remain, use nodewise recovery.
```
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Note to add nodewise recovery doc when available

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 2]
```diff
@@ -0,0 +1,30 @@
+= Partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can arise when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Partition recovery is carried out by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. However, this allows some potential data loss on the partition if the best available replica is out of sync, resulting in an unclean leader election. Partition recovery should only be used when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority.
```
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Please verify how RP picks the "best available" replica

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+If you want to instead permanently remove failed brokers from a cluster, especially in situations where no surviving partition replicas remain, use nodewise recovery.  
+
+== Use the Admin API to recover a partition
+
+Make a POST request to the `/force_replicas` endpoint:
+
+[,bash]
+----
+curl -X POST http://localhost:9644/v1/debug/partitions/<namespace>/<topic>/<partition>/force_replicas \
```
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Can we provide a better, working example?

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  -d '{
+    [
+        { "core": 0, "node_id": 0 } 
+    ]
+  }'
```
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc:75]: The request body schema requires an array of replica assignments. What is the definition of `core` and how do we find its values?

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 2 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: lgtm.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 1 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: > instead permanently remove failed brokers from a cluster

I'd rephrase this. node wise recovery is just bulk partition force recovery, it doesn't remove failed brokers. Instead of doing force recovery manually for every partition on the failed broker, node wise recovery provides a way to to perform that action in one go.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 5]
```diff
@@ -0,0 +1,30 @@
+= Partition recovery
```
Comment (bharathv) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Forced partition recovery.

IMO force signifies that it is not a regular raft recovery.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 4 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc:75]: core is the CPU core aka shard id. If there are n CPU cores on the machine, user has to pick a value in range [0, n-1], currently user has to pick a random value or find a least loaded shard (if they know what that is) and use that value.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 5 above)
Comment (kbatuigas) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: I can change it, @mattschumpert any thoughts?

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can arise when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader.
```
Comment (Feediver1) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]:
```suggestion
You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can occur when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
```

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 1 above)
Comment (Feediver1) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: @greg-szrama should be reviewing this too.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+----
+curl -X POST http://localhost:9644/v1/debug/partitions/<namespace>/<topic>/<partition>/force_replicas \
+  -d '{
+    [
+        { "core": 0, "node_id": 0 } 
+    ]
+  }'
+----
+
+See xref:api:ROOT:admin-api.adoc#tag/Partitions/operation/force_update_partition_replicas[Redpanda Admin API] for a full reference.
```
Comment (Feediver1) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]:
```suggestion
See xref:api:ROOT:admin-api.adoc#tag/Partitions/operation/force_update_partition_replicas[Redpanda Admin API] for additional detail.
```

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 5 above)
Comment (mattschumpert) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Agreed

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 8]
```diff
@@ -0,0 +1,31 @@
+= Partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can occur when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader.
```
Comment (mattschumpert) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: is 'appears to be stuck' the best way to describe it @bharathv ?

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 4 above)
Comment (mattschumpert) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc:75]: @kbatuigas @bharathv this doesn't talk about specifying the actual partition ID which seems to be the most important thing!

I get that you have to specify the node and core (which seems awkward in itself), but what about the specific partitions.  We need to give an example with the actual partition ID you're using.

**Review by mattschumpert — CHANGES_REQUESTED** [review]

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 4 above)
Comment (kbatuigas) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc:75]: @mattschumpert Looks like the partition ID is specified in the request URL `/debug/partitions/{namespace}/{topic}/{partition}/force_replicas`. @bharathv can you please provide example values for the following?
- `namespace`
- `topic`
- `partition` 
- `node_id`

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 3 above)
Comment (greg-szrama) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Is this curl only or is there also an RPK command?

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 9]
```diff
@@ -0,0 +1,31 @@
+= Forced partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can occur when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Forced partition recovery is carried out by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. However, this allows some potential data loss on the partition if the best available replica is out of sync, resulting in an unclean leader election. Forced partition recovery should only be used when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
+
+If you want to instead force recover partitions in bulk on a failed broker, use nodewise recovery.
```
Comment (greg-szrama) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: I'd rephrase this -- it's not so much when you want to force recover partitions as needing to do a recovery with whole dead nodes. Nodewise partition recovery force-recovers a node, letting partitions which have a majority be rebuilt while starting fresh partitions where there is no majority.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 8 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: I think "leaderless because of lost majority of replicas" I think is the right way (please rephrase as you see fit).

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 9 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: > Nodewise partition recovery force-recovers a node

Just to be clear it doesn't recover the node itself (perhaps you meant the same thing but just want to call out incase it is not obvious to someone), it just rebuilds the partition replicas on those dead nodes elsewhere on other available brokers.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 3 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: there is no rpk command that I'm aware of

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 4 above)
Comment (bharathv) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc:75]: Right, partition ID is in the URL,


Here is a working example, lets assume partition test/0 is leaderless because majority of replicas are gone. 

A user is expected to get the current node/shard assignments using the following command

```
curl http://$(hostname):9644/v1/partitions/kafka/test | jq

[
  {
    "ns": "kafka",
    "topic": "test",
    "partition_id": 0,
    "status": "done",
    "leader_id": 1,
    "raft_group_id": 1,
    "replicas": [
      {
        "node_id": 1,
        "core": 1
      },
      {
        "node_id": 3,
        "core": 1
      },
      {
        "node_id": 2,
        "core": 1
      }
    ]
  }
]
```


Lets assume nodes 2, 3 are dead and we want to replace them with nodes 4, 5, the command would look like this.


```
curl -X POST http://$(hostname):9644/v1/partitions/kafka/test/0/force_replicas  -H 'Content-Type: application/json' --data '[{"node_id": 1, "core": 1}, {"node_id": 4, "core": 1}, {"node_id": 5, "core": 0}]'
```

As you can see the input to the command is the new replica set that will force replace the current one. It includes {1, 1} (which ensures we recover data from it) and two new replica assignments {4, 1}, {5, 0} which will hydrate data from {1, 1}.

User has to pick least loaded cores and the system may later move around the replicas if they are imbalanced.

Clearly this a low level operation and has to be done with caution.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 10]
```diff
@@ -0,0 +1,31 @@
+= Forced partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can occur when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Forced partition recovery is carried out by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. However, this allows some potential data loss on the partition if the best available replica is out of sync, resulting in an unclean leader election. Forced partition recovery should only be used when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority.
```
Comment (greg-szrama) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]:
```suggestion
Redpanda performs forced partition recovery by promoting the best available replica to leader, thereby making that partition available for produce and consume. The new leader is typically the replica with the highest offset. However, this allows some potential data loss on the partition if the best available replica is out of sync, resulting in an unclean leader election. Forced partition recovery should only be used when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
```

You have some passive voice in here, you might rephrase it a bit.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+== Use the Admin API to recover a partition
+
+The following examples assume that partition 0 in topic `test` is unavailable and its replicas cannot form a majority to elect a leader.
+
+. Call the xref:api:ROOT:admin-api.adoc#tag/Partitions/operation/get_topic_partitions[`/partitions/kafka/<topic>`] endpoint to determine the current broker and shard assignments of the partition replicas.
++
+[,bash]
+----
+curl http://localhost:9644/v1/partitions/kafka/test | jq
```
Comment (bharathv) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: nit: you can remove the `| jq` part, I used it to format the json

actual command is just

```
curl http://localhost:9644/v1/partitions/kafka/test
```

**Review by bharathv — APPROVED** [review]

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+Redpanda performs forced partition recovery by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. 
+
+CAUTION: Forced partition recovery allows some potential data loss on the partition if the best available replica is out of sync, ending up in an unclean leader election. Use this operation with caution, and only when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
+
+NOTE: If you want to instead force recover all partitions in bulk from a set of failed brokers, use nodewise recovery.
+
+== Use the Admin API to recover a partition
+
+The following examples assume that partition 0 in topic `test` is unavailable and its replicas cannot form a majority to elect a leader.
```
Comment (Feediver1) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc:14]:
```suggestion
The following examples are based on the assumption that partition 0 in topic `test` is unavailable and its replicas cannot form a majority to elect a leader.
```

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+      {
+        "node_id": 3,
+        "core": 1
+      }
+    ]
+  }
+]
+----
+
+. Suppose brokers 2 and 3 have failed and we want to replace them with brokers 4 and 5.
```
Comment (Feediver1) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]:
```suggestion
. In this scenario, brokers 2 and 3 have failed and you want to replace them with brokers 4 and 5.
```

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 14]
```diff
@@ -0,0 +1,86 @@
+= Forced partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that is unavailable and has lost a majority of its replicas. This can occur when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Redpanda performs forced partition recovery by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. 
+
+CAUTION: Forced partition recovery allows some potential data loss on the partition if the best available replica is out of sync, ending up in an unclean leader election. Use this operation with caution, and only when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
+
+NOTE: If you want to instead force recover all partitions in bulk from a set of failed brokers, use nodewise recovery.
```
Comment (Feediver1) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: Can you provide a link to the instructions for nodewise recovery?

**Review by Feediver1 — APPROVED** [review]
See updates in context

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 14 above)
Comment (kbatuigas) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: This will be added when the doc is available cc @greg-szrama

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 15]
```diff
@@ -0,0 +1,86 @@
+= Forced partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that is unavailable and has lost a majority of its replicas. This can occur when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Redpanda performs forced partition recovery by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset.
```
Comment (greg-szrama) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]:
```suggestion
Redpanda performs forced partition recovery by promoting the best available replica to leader, making the partition available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. 
```

Passive voice here

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+CAUTION: Forced partition recovery allows some potential data loss on the partition if the best available replica is out of sync, ending up in an unclean leader election. Use this operation with caution, and only when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
+
+NOTE: If you want to instead force recover all partitions in bulk from a set of failed brokers, use xref:./nodewise-partition-recovery.adoc[nodewise recovery].
+
+== Use the Admin API to recover a partition
+
+The following examples are based on the assumption that partition 0 in topic `test` is unavailable and its replicas cannot form a majority to elect a leader.
+
+. Call the xref:api:ROOT:admin-api.adoc#tag/Partitions/operation/get_topic_partitions[`/partitions/kafka/<topic>`] endpoint to determine the current broker and shard assignments of the partition replicas.
```
Comment (greg-szrama) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc:16]: The example here is good, I'm just checking whether there's also an RPK command to do this. If so, let's reference it in this section as well.

**Review by greg-szrama — APPROVED** [review]
Looks good overall! Just one question in there and a small edit.

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc (hunk 16 above)
Comment (mattschumpert) [reply modules/manage/pages/cluster-maintenance/partition-recovery.adoc:16]: I don't think so. @ztlpn ?

File: modules/manage/pages/cluster-maintenance/partition-recovery.adoc [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+      {
+        "node_id": 3,
+        "core": 1
+      }
+    ]
+  }
+]
+----
+
+. In this scenario, brokers 2 and 3 have failed and you want to replace them with brokers 4 and 5.
```
Comment (mattschumpert) [inline modules/manage/pages/cluster-maintenance/partition-recovery.adoc]: @kbatuigas maybe instead of 'replace them', 'we want to move the replicas from those nodes to brokers 4 and 5, which are healthy'

**Review by mattschumpert — APPROVED** [review]

**Review by mmaslankaprv — APPROVED** [review]
